lib/Epub/EpubList/Epub.cpp

```diff
@@ -177,6 +177,76 @@ bool Epub::load()
   return true;
 }
 
+// load in the TOC index data for the epub file (Chapters)
+bool Epub::loadIndex()
+{
+  ZipFile zip(m_path.c_str());
+  // Update this so it's read from manifest v
+  // Or alternatively reads an ncx file in OEBPS directory
+  ESP_LOGI(TAG, "toc path: %s\n", get_toc_filename().c_str());
+  const char *tocFile = get_toc_filename().c_str();
+  char *meta_index = (char *)zip.read_file_to_memory(tocFile);
+  if (!meta_index)
+  {
+    ESP_LOGE(TAG, "Could not find %s", tocFile);
+    return false;
+  }
+  // Parse the Toc contents
+  tinyxml2::XMLDocument doc;
+  auto result = doc.Parse(meta_index);
+  free(meta_index);
+  if (result != tinyxml2::XML_SUCCESS)
+  {
+    ESP_LOGE(TAG, "Error parsing toc %s", doc.ErrorIDToName(result));
+    return false;
+  }
+  auto ncx = doc.FirstChildElement("ncx");
+  if (!ncx)
+  {
+    ESP_LOGE(TAG, "Could not find first child ncx in toc");
+    return false;
+  }
+  printf("ncx in line:%d\n", ncx->GetLineNum());
+
+  auto docTitle = ncx->FirstChildElement("docTitle");
+  if (!docTitle)
+  {
+    ESP_LOGE(TAG, "Could not find docTitle child in ncx");
+    return false;
+  }
+
+  auto navMap = ncx->FirstChildElement("navMap");
+  if (!navMap)
+  {
+    ESP_LOGE(TAG, "Could not find navMap child in ncx");
+    return false;
+  }
+
+  auto navPoint = navMap->FirstChildElement("navPoint");
+
+  // Fills toc_index map
+  while (navPoint)
+  {
+    // navPoint has also an id & playOrder element: navPoint->Attribute("id");
+    auto navLabel = navPoint->FirstChildElement("navLabel")->FirstChildElement("text")->FirstChild();
+    std::string title = navLabel->Value();
+    auto content = navPoint->FirstChildElement("content");
+    std::string href = content->Attribute("src");
+
```

Comment:
I see two options here:

1.  We add a second iterator. In case a nested navPoint is found it just adds more indexes with just two spaces at the left
2. We add a 3rd uint8_t **deep_level** to the vector to signalize that. And we use that to layout it differently (Maybe too much?)

Reply:
https://github.com/atomic14/diy-esp32-epub-reader/issues/70